File: modules/guardian-node/jaeger/threat_hint_v2_proof_envelope.py
"""Local-only canonical ThreatHint v2 proof envelope parsing and binding."""

# Exact built-in types and a fixed six-field wire are protocol requirements.
# pylint: disable=too-many-instance-attributes,unidiomatic-typecheck

from __future__ import annotations

import json
import re
import weakref
from dataclasses import dataclass
from typing import Any, Dict, Iterable

from jaeger.threat_hint_v2_statement import (
    ThreatHintV2Statement,
    ThreatHintV2StatementError,
)

MAX_CANONICAL_V2_PROOF_ENVELOPE_BYTES = 4096
SCHEMA_VERSION = 2
PROTOCOL_ID = "/prometheus/threat-hint/2.0.0"
RELATION_ID = "prometheus-threat-hint-v2"
MIN_PROOF_BYTES = 1
MAX_PROOF_BYTES = 1024


class ThreatHintV2ProofEnvelopeError(ValueError):
    """Redacted local parser/validator failure."""

    def __init__(self) -> None:
        super().__init__("invalid threat-hint v2 proof envelope")


@dataclass(frozen=True, init=False, repr=False, eq=False)
class ThreatHintV2ProofEnvelope:
    """Canonical local v2 proof envelope with no proof or transport authority.

    Structural validity is not proof acceptance: the proof bytes are opaque,
    no proof system is named here, and this type grants no authority against
    arbitrary in-process code. A later RelationManifest v2 binds the relation
    to a separately approved proof system and keys.
    """

    schema_version: int
    protocol_id: str
    relation_id: str
    statement: str
    statement_digest: str
    proof: str
    parsed_statement: ThreatHintV2Statement

    def __init__(self) -> None:
        """Reject direct construction outside canonical parsing."""
        raise TypeError("direct threat-hint v2 proof envelope construction is disabled")

    @classmethod
    def parse_canonical(
        cls, wire_bytes: bytes, trusted_network_id: str
    ) -> "ThreatHintV2ProofEnvelope":
        """Parse exact bytes against a separately trusted local network."""
        if cls is not ThreatHintV2ProofEnvelope or type(wire_bytes) is not bytes:
            raise ThreatHintV2ProofEnvelopeError()
        if (
            len(wire_bytes) == 0
            or len(wire_bytes) > MAX_CANONICAL_V2_PROOF_ENVELOPE_BYTES
        ):
            raise ThreatHintV2ProofEnvelopeError()
        if type(trusted_network_id) is not str:
            raise ThreatHintV2ProofEnvelopeError()

        try:
            decoded = json.loads(
                wire_bytes.decode("utf-8"), object_pairs_hook=_reject_duplicate_keys
            )
        except (UnicodeDecodeError, ValueError):
            raise ThreatHintV2ProofEnvelopeError() from None

        envelope = cls._parse_object(decoded, trusted_network_id)
        _PARSED_CANONICAL[envelope] = wire_bytes
        if envelope.canonical_bytes != wire_bytes:
            raise ThreatHintV2ProofEnvelopeError()
        return envelope

    @classmethod
    def _parse_object(
        cls, value: Any, trusted_network_id: str
    ) -> "ThreatHintV2ProofEnvelope":
        """Validate one decoded exact-shape envelope object."""
        if type(value) is not dict or list(value.keys()) != [
            "schema_version",
            "protocol_id",
            "relation_id",
            "statement",
            "statement_digest",
            "proof",
        ]:
            raise ThreatHintV2ProofEnvelopeError()

        schema_version = _exact_schema_version(value["schema_version"])
        protocol_id = _exact_identifier(value["protocol_id"], PROTOCOL_ID)
        relation_id = _exact_identifier(value["relation_id"], RELATION_ID)
        statement_digest = _fixed_lower_hex(value["statement_digest"])
        proof = _proof_hex(value["proof"])

        statement = value["statement"]
        if type(statement) is not str:
            raise ThreatHintV2ProofEnvelopeError()
        try:
            parsed_statement = ThreatHintV2Statement.parse_canonical(
                statement.encode("utf-8"), trusted_network_id
            )
        except ThreatHintV2StatementError:
            raise ThreatHintV2ProofEnvelopeError() from None
        if parsed_statement.statement_digest().hex() != statement_digest:
            raise ThreatHintV2ProofEnvelopeError()

        envelope = object.__new__(cls)
        object.__setattr__(envelope, "schema_version", schema_version)
        object.__setattr__(envelope, "protocol_id", protocol_id)
        object.__setattr__(envelope, "relation_id", relation_id)
        object.__setattr__(envelope, "statement", statement)
        object.__setattr__(envelope, "statement_digest", statement_digest)
        object.__setattr__(envelope, "proof", proof)
        object.__setattr__(envelope, "parsed_statement", parsed_statement)
        return envelope

    @property
    def canonical_bytes(self) -> bytes:
        """Return the revalidated exact canonical JSON bytes."""
        canonical_at_parse = _PARSED_CANONICAL.get(self)
        if canonical_at_parse is None:
            raise ThreatHintV2ProofEnvelopeError()
        self._validate_state()
        payload = {
            "schema_version": self.schema_version,
            "protocol_id": self.protocol_id,
            "relation_id": self.relation_id,
            "statement": self.statement,
            "statement_digest": self.statement_digest,
            "proof": self.proof,
        }
        canonical = json.dumps(
            payload, separators=(",", ":"), ensure_ascii=False
        ).encode("utf-8")
        if (
            len(canonical) == 0
            or len(canonical) > MAX_CANONICAL_V2_PROOF_ENVELOPE_BYTES
        ):
            raise ThreatHintV2ProofEnvelopeError()
        if canonical != canonical_at_parse:
            raise ThreatHintV2ProofEnvelopeError()
        return canonical

    def proof_bytes(self) -> bytes:
        """Return the opaque proof bytes without interpreting them."""
        if _PARSED_CANONICAL.get(self) is None:
            raise ThreatHintV2ProofEnvelopeError()
        self._validate_state()
        return bytes.fromhex(self.proof)

    def _validate_state(self) -> None:
        """Revalidate private state before serialization or decoding."""
        if type(self) is not ThreatHintV2ProofEnvelope:
            raise ThreatHintV2ProofEnvelopeError()
        _exact_schema_version(self.schema_version)
        _exact_identifier(self.protocol_id, PROTOCOL_ID)
        _exact_identifier(self.relation_id, RELATION_ID)
        _fixed_lower_hex(self.statement_digest)
        _proof_hex(self.proof)
        if type(self.statement) is not str:
            raise ThreatHintV2ProofEnvelopeError()
        if type(self.parsed_statement) is not ThreatHintV2Statement:
            raise ThreatHintV2ProofEnvelopeError()
        if self.parsed_statement.canonical_bytes != self.statement.encode("utf-8"):
            raise ThreatHintV2ProofEnvelopeError()
        if self.parsed_statement.statement_digest().hex() != self.statement_digest:
            raise ThreatHintV2ProofEnvelopeError()


def _reject_duplicate_keys(items: Iterable[tuple[str, Any]]) -> Dict[str, Any]:
    result: Dict[str, Any] = {}
    for key, value in items:
        if key in result:
            raise ThreatHintV2ProofEnvelopeError()
        result[key] = value
    return result


def _exact_schema_version(value: Any) -> int:
    if type(value) is not int or value != SCHEMA_VERSION:
        raise ThreatHintV2ProofEnvelopeError()
    return value


def _exact_identifier(value: Any, expected: str) -> str:
    if type(value) is not str or value != expected:
        raise ThreatHintV2ProofEnvelopeError()
    return value


def _fixed_lower_hex(value: Any) -> str:
    if (
        type(value) is not str
        or len(value) != 64
        or _LOWER_HEX_RE.fullmatch(value) is None
    ):
        raise ThreatHintV2ProofEnvelopeError()
    return value


def _proof_hex(value: Any) -> str:
    if (
        type(value) is not str
        or len(value) % 2 != 0
        or _PROOF_HEX_RE.fullmatch(value) is None
    ):
        raise ThreatHintV2ProofEnvelopeError()
    if not MIN_PROOF_BYTES <= len(value) // 2 <= MAX_PROOF_BYTES:
        raise ThreatHintV2ProofEnvelopeError()
    return value


_LOWER_HEX_RE = re.compile(r"[0-9a-f]{64}")
_PROOF_HEX_RE = re.compile(r"[0-9a-f]*")
_PARSED_CANONICAL: weakref.WeakKeyDictionary[Any, bytes] = weakref.WeakKeyDictionary()

File: modules/threat-hint/src/threat_hint_v2_proof_envelope.rs
//! Local-only canonical ThreatHint v2 proof envelope parsing and binding.
//!
//! Structural validity is not proof acceptance: this slice never verifies the
//! proof bytes and never grants any authority. There is deliberately no
//! `proof_system` field and no development stub; a later RelationManifest v2
//! binds `relation_id` to a separately approved proof system and keys.

use serde::{Deserialize, Serialize};
use thiserror::Error;

use crate::observable_bundle::validate_network_id;
use crate::threat_hint_v2_statement::ThreatHintV2Statement;
use crate::MAX_PROOF_BYTES;

const PROOF_ENVELOPE_SCHEMA_VERSION: u16 = 2;
const MIN_PROOF_BYTES: usize = 1;
const DIGEST_HEX_LEN: usize = 64;

/// The only supported ThreatHint v2 proof envelope protocol identifier.
pub const THREAT_HINT_V2_PROTOCOL_ID: &str = "/prometheus/threat-hint/2.0.0";

/// The only supported ThreatHint v2 proof envelope relation identifier.
pub const THREAT_HINT_V2_RELATION_ID: &str = "prometheus-threat-hint-v2";

/// Maximum accepted canonical ThreatHint v2 proof envelope size.
pub const MAX_CANONICAL_V2_PROOF_ENVELOPE_BYTES: usize = 4096;

/// Redacted failure returned for every invalid envelope or trusted network.
#[derive(Debug, Error, PartialEq, Eq)]
pub enum ThreatHintV2ProofEnvelopeError {
    /// The envelope or separately trusted network is invalid.
    #[error("invalid threat-hint v2 proof envelope")]
    InvalidEnvelope,
}

/// A canonical local ThreatHint v2 proof envelope.
///
/// The envelope binds a canonical GH-114 statement and its domain-separated
/// digest to opaque proof bytes. Structural validity is not proof acceptance:
/// no proof system is named, and the proof bytes are never verified here.
///
/// Direct deserialization is deliberately unavailable; callers must use
/// [`ThreatHintV2ProofEnvelope::parse_canonical`] with a separately trusted
/// network.
///
/// ```compile_fail
/// use prometheus_threat_hint::ThreatHintV2ProofEnvelope;
///
/// let _: ThreatHintV2ProofEnvelope = serde_json::from_slice(b"{}").unwrap();
/// ```
#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
pub struct ThreatHintV2ProofEnvelope {
    schema_version: u16,
    protocol_id: String,
    relation_id: String,
    statement: String,
    statement_digest: String,
    proof: String,
    #[serde(skip)]
    parsed_statement: ThreatHintV2Statement,
}

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct ThreatHintV2ProofEnvelopeWire {
    schema_version: u16,
    protocol_id: String,
    relation_id: String,
    statement: String,
    statement_digest: String,
    proof: String,
}

impl ThreatHintV2ProofEnvelope {
    /// The only supported ThreatHint v2 proof envelope schema version.
    pub const SCHEMA_VERSION: u16 = PROOF_ENVELOPE_SCHEMA_VERSION;

    /// Parses exact canonical bytes against a separately trusted local network.
    pub fn parse_canonical(
        wire_bytes: &[u8],
        trusted_network_id: &str,
    ) -> Result<Self, ThreatHintV2ProofEnvelopeError> {
        if wire_bytes.is_empty() || wire_bytes.len() > MAX_CANONICAL_V2_PROOF_ENVELOPE_BYTES {
            return Err(ThreatHintV2ProofEnvelopeError::InvalidEnvelope);
        }
        validate_network_id(trusted_network_id)
            .map_err(|_| ThreatHintV2ProofEnvelopeError::InvalidEnvelope)?;

        let wire: ThreatHintV2ProofEnvelopeWire = serde_json::from_slice(wire_bytes)
            .map_err(|_| ThreatHintV2ProofEnvelopeError::InvalidEnvelope)?;
        let parsed_statement =
            ThreatHintV2Statement::parse_canonical(wire.statement.as_bytes(), trusted_network_id)
                .map_err(|_| ThreatHintV2ProofEnvelopeError::InvalidEnvelope)?;

        let envelope = Self {
            schema_version: wire.schema_version,
            protocol_id: wire.protocol_id,
            relation_id: wire.relation_id,
            statement: wire.statement,
            statement_digest: wire.statement_digest,
            proof: wire.proof,
            parsed_statement,
        };

        envelope.validate()?;
        envelope.validate_statement_binding()?;
        if envelope.to_canonical_bytes()? != wire_bytes {
            return Err(ThreatHintV2ProofEnvelopeError::InvalidEnvelope);
        }

        Ok(envelope)
    }

    /// Returns the parsed schema version.
    pub fn schema_version(&self) -> u16 {
        self.schema_version
    }

    /// Returns the exact protocol identifier.
    pub fn protocol_id(&self) -> &str {
        &self.protocol_id
    }

    /// Returns the exact relation identifier.
    pub fn relation_id(&self) -> &str {
        &self.relation_id
    }

    /// Returns the bound canonical GH-114 statement.
    pub fn statement(&self) -> &ThreatHintV2Statement {
        &self.parsed_statement
    }

    /// Returns the bound statement digest as lowercase hexadecimal.
    pub fn statement_digest_hex(&self) -> &str {
        &self.statement_digest
    }

    /// Returns the opaque proof bytes as lowercase hexadecimal.
    pub fn proof_hex(&self) -> &str {
        &self.proof
    }

    /// Decodes the opaque proof bytes without interpreting them.
    pub fn proof_bytes(&self) -> Result<Vec<u8>, ThreatHintV2ProofEnvelopeError> {
        if !is_valid_proof_hex(&self.proof) {
            return Err(ThreatHintV2ProofEnvelopeError::InvalidEnvelope);
        }
        hex::decode(&self.proof).map_err(|_| ThreatHintV2ProofEnvelopeError::InvalidEnvelope)
    }

    /// Serializes the validated envelope to exact canonical JSON bytes.
    pub fn to_canonical_bytes(&self) -> Result<Vec<u8>, ThreatHintV2ProofEnvelopeError> {
        self.validate()?;
        self.validate_statement_binding()?;
        let bytes = serde_json::to_vec(self)
            .map_err(|_| ThreatHintV2ProofEnvelopeError::InvalidEnvelope)?;
        if bytes.is_empty() || bytes.len() > MAX_CANONICAL_V2_PROOF_ENVELOPE_BYTES {
            return Err(ThreatHintV2ProofEnvelopeError::InvalidEnvelope);
        }
        Ok(bytes)
    }

    fn validate(&self) -> Result<(), ThreatHintV2ProofEnvelopeError> {
        if self.schema_version != Self::SCHEMA_VERSION
            || self.protocol_id != THREAT_HINT_V2_PROTOCOL_ID
            || self.relation_id != THREAT_HINT_V2_RELATION_ID
            || !is_lower_hex_64(&self.statement_digest)
            || !is_valid_proof_hex(&self.proof)
        {
            return Err(ThreatHintV2ProofEnvelopeError::InvalidEnvelope);
        }
        Ok(())
    }

    fn validate_statement_binding(&self) -> Result<(), ThreatHintV2ProofEnvelopeError> {
        let statement_bytes = self
            .parsed_statement
            .to_canonical_bytes()
            .map_err(|_| ThreatHintV2ProofEnvelopeError::InvalidEnvelope)?;
        if self.statement.as_bytes() != statement_bytes.as_slice() {
            return Err(ThreatHintV2ProofEnvelopeError::InvalidEnvelope);
        }
        let digest = self
            .parsed_statement
            .statement_digest()
            .map_err(|_| ThreatHintV2ProofEnvelopeError::InvalidEnvelope)?;
        if hex::encode(digest) != self.statement_digest {
            return Err(ThreatHintV2ProofEnvelopeError::InvalidEnvelope);
        }
        Ok(())
    }
}

fn is_lower_hex_64(value: &str) -> bool {
    value.len() == DIGEST_HEX_LEN
        && value
            .as_bytes()
            .iter()
            .all(|byte| matches!(byte, b'0'..=b'9' | b'a'..=b'f'))
}

fn is_valid_proof_hex(value: &str) -> bool {
    if !value.len().is_multiple_of(2)
        || !value
            .as_bytes()
            .iter()
            .all(|byte| matches!(byte, b'0'..=b'9' | b'a'..=b'f'))
    {
        return false;
    }
    let raw_len = value.len() / 2;
    (MIN_PROOF_BYTES..=MAX_PROOF_BYTES).contains(&raw_len)
}

#[cfg(test)]
mod tests {
    use super::*;

    const STATEMENT_WIRE: &[u8] = br#"{"schema_version":2,"artifact_hash":"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa","observable_commitment":"bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb","confidence_bps":7500,"disclosure_class":"review_required_v1","report_nonce":"cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc","observed_at":1700000000,"network_id":"testnet-10"}"#;

    fn sample_wire() -> Vec<u8> {
        let statement =
            ThreatHintV2Statement::parse_canonical(STATEMENT_WIRE, "testnet-10").expect("valid");
        let digest = hex::encode(statement.statement_digest().expect("digest"));
        let statement_text = String::from_utf8(STATEMENT_WIRE.to_vec()).expect("ASCII");
        format!(
            "{{\"schema_version\":2,\"protocol_id\":\"{THREAT_HINT_V2_PROTOCOL_ID}\",\"relation_id\":\"{THREAT_HINT_V2_RELATION_ID}\",\"statement\":{},\"statement_digest\":\"{digest}\",\"proof\":\"{}\"}}",
            serde_json::to_string(&statement_text).expect("string"),
            "aa".repeat(16)
        )
        .into_bytes()
    }

    #[test]
    fn roundtrip_byte_identity_and_binding() {
        let wire = sample_wire();
        let envelope =
            ThreatHintV2ProofEnvelope::parse_canonical(&wire, "testnet-10").expect("valid");

        assert_eq!(envelope.schema_version(), 2);
        assert_eq!(envelope.protocol_id(), THREAT_HINT_V2_PROTOCOL_ID);
        assert_eq!(envelope.relation_id(), THREAT_HINT_V2_RELATION_ID);
        assert_eq!(
            envelope
                .statement()
                .to_canonical_bytes()
                .expect("canonical"),
            STATEMENT_WIRE
        );
        assert_eq!(
            envelope.statement_digest_hex(),
            hex::encode(envelope.statement().statement_digest().expect("digest"))
        );
        assert_eq!(envelope.proof_bytes().expect("proof"), vec![0xaa; 16]);
        assert_eq!(envelope.to_canonical_bytes().expect("canonical"), wire);
    }

    #[test]
    fn rejects_empty_oversized_and_untrusted_network() {
        assert_eq!(
            ThreatHintV2ProofEnvelope::parse_canonical(b"", "testnet-10"),
            Err(ThreatHintV2ProofEnvelopeError::InvalidEnvelope)
        );
        assert_eq!(
            ThreatHintV2ProofEnvelope::parse_canonical(
                &vec![b'{'; MAX_CANONICAL_V2_PROOF_ENVELOPE_BYTES + 1],
                "testnet-10"
            ),
            Err(ThreatHintV2ProofEnvelopeError::InvalidEnvelope)
        );
        assert_eq!(
            ThreatHintV2ProofEnvelope::parse_canonical(&sample_wire(), "mainnet"),
            Err(ThreatHintV2ProofEnvelopeError::InvalidEnvelope)
        );
        assert_eq!(
            ThreatHintV2ProofEnvelope::parse_canonical(&sample_wire(), "-testnet-10"),
            Err(ThreatHintV2ProofEnvelopeError::InvalidEnvelope)
        );
    }

    #[test]
    fn rejects_trailing_and_noncanonical_input() {
        let mut trailing = sample_wire();
        trailing.push(b'\n');
        assert_eq!(
            ThreatHintV2ProofEnvelope::parse_canonical(&trailing, "testnet-10"),
            Err(ThreatHintV2ProofEnvelopeError::InvalidEnvelope)
        );

        let spaced = String::from_utf8(sample_wire())
            .expect("ASCII")
            .replace(",\"protocol_id\"", ", \"protocol_id\"");
        assert_eq!(
            ThreatHintV2ProofEnvelope::parse_canonical(spaced.as_bytes(), "testnet-10"),
            Err(ThreatHintV2ProofEnvelopeError::InvalidEnvelope)
        );
    }
}
